tensorflow/core/kernels/svd_op_gpu.cu.cc

```diff
@@ -0,0 +1,249 @@
+                         done);
+    OP_REQUIRES_OK_ASYNC(context, context->allocate_output(2, shapeV, &outputV),
+                         done);
+    if (compute_uv_) {
+      OP_REQUIRES_OK_ASYNC(
+          context, context->allocate_temp(input.dtype(), shapeVT, &outputVT),
+          done);
+    }
+
+    if (n == 0 || m == 0) {
+      // If X is an empty matrix (0 rows, 0 col), X * X' == X.
+      // Therefore, we return X.
+      done();
+      return;
+    }
+
+    // Copy and reshape input tensor
+    // SVD modifies the input, so I need to copy it.
+    Tensor inputCopy;
+    OP_REQUIRES_OK_ASYNC(
+        context,
+        context->allocate_temp(input.dtype(), input.shape(), &inputCopy), done);
+
+    cudaMemcpy(inputCopy.flat<Scalar>().data(), input.flat<Scalar>().data(),
+               input.NumElements() * sizeof(Scalar), cudaMemcpyDeviceToDevice);
+
+    auto input_reshaped = inputCopy.template flat_inner_dims<Scalar, 3>();
+
+    // Reshape output tensors
+    auto outputS_reshaped = outputS->template flat_inner_dims<Scalar, 2>();
+    Scalar* outputU_reshaped_ptr = NULL;
+    Scalar* outputVT_reshaped_ptr = NULL;
+    if (compute_uv_) {
+      auto outputU_reshaped = outputU->template flat_inner_dims<Scalar, 3>();
+      auto outputVT_reshaped = outputVT.template flat_inner_dims<Scalar, 3>();
+      outputU_reshaped_ptr = outputU_reshaped.data();
+      outputVT_reshaped_ptr = outputVT_reshaped.data();
+    }
+
+    // Launch a SVD kernel for each matrix in the batch.
+    const int64 batch_size = input_reshaped.dimension(0);
+    DeviceLapackInfo dev_info(context, batch_size, "gesvd");
+
+    // TODO(rmlarsen): Parallelize over batches if it turns out to be
+    // an important use case.
+    CudaSolver solver(context);
+    for (int64 i = 0; i < batch_size; ++i) {
+      Scalar* input_ptr = input_reshaped.data() + i * m * n;
+      Scalar* outputS_ptr = outputS_reshaped.data() + i * p;
+      int lda = m;
+
+      Scalar* outputU_ptr = NULL;
+      Scalar* outputVT_ptr = NULL;
+      signed char jobu = 'N';
+      signed char jobvt = 'N';
+      int ldu = m;
+      int ldvt = n;
+      if (compute_uv_) {
+        if (full_matrices_) {
+          outputU_ptr = outputU_reshaped_ptr + i * m * m;
+          outputVT_ptr = outputVT_reshaped_ptr + i * n * n;
+          jobu = 'A';
+          jobvt = 'A';
+        } else {
+          outputU_ptr = outputU_reshaped_ptr + i * m * p;
+          outputVT_ptr = outputVT_reshaped_ptr + i * n * p;
+          jobu = 'S';
+          jobvt = 'S';
+        }
+      }
+
+      int* dev_info_ptr = dev_info.mutable_data() + i;
+      OP_REQUIRES_OK_ASYNC(
+          context,
+          solver.Gesvd(jobu, jobvt, m, n, input_ptr, lda, outputS_ptr,
+                       outputU_ptr, ldu, outputVT_ptr, ldvt, dev_info_ptr),
+          done);
+    }
+
+    // Test if it was successfull.
+    // I'm not using solver.CopyLapackInfoToHostAsync
```

Comment:
Alternatively, you can just launch the SVD and the subsequent transpose kernel opportunistically up front, and just use solver.CopyLapackInfoToHostAsync to check for the status at the end. The SVD and transpose kernels will get serialized on the stream.